### ISSUE TITLE
"write after end" crashes that happen on multiple clients simultaneously

### ISSUE THREAD

**Issue by Lange** [twitch-irc/twitch-irc#83]
This bug is fairly low-priority for me as my production setup can recover from it with no issue, but it seemed worth reporting.

Occasionally, one or more of my twitch-irc instances will simultaneously crash with the following error:

```
2015-04-23 22:32: error: [lfg-sublistener] CRASH: write after end

 Error: write after end
    at writeAfterEnd (/nodecg_directory/bundles/lfg-sublistener/node_modules/twitch-irc/node_modules/readable-stream/lib/_stream_writable.js:144:12)
    at DestroyableTransform.Writable.write (/nodecg_directory/bundles/lfg-sublistener/node_modules/twitch-irc/node_modules/readable-stream/lib/_stream_writable.js:192:5)
    at Socket.ondata (_stream_readable.js:520:20)
    at emitOne (events.js:82:20)
    at Socket.emit (events.js:166:7)
    at readableAddChunk (_stream_readable.js:145:16)
    at Socket.Readable.push (_stream_readable.js:109:10)
    at TCP.onread (net.js:508:20)
```

I'm assuming this is somehow caused by something received from twitch. What's strange is that it doesn't happen to all of my instances, only some of them. Perhaps it's somehow dependent on some kind of tick cycle.

**Comment by Maxorq** [comment]
that's the error i get when the module tries to reconnect after a disconnection.
reconnecting is pretty much useless (not working at all?), easier to just crash the whole app and start it over

**Comment by OurFlagIsMined** [comment]
**_TL;DR skip to the bold section for the one-line quick fix to bypass the fastReconnect issue_**

I've been having the same problem; periodic crashes with the same 'write after end' error, always following a reconnect request from TMI (though not immediately after; the last one I saw didn't crash until 255 seconds after the old connection was dropped, and sometimes it doesn't crash at all, though other problems occur in those cases where it does survive). When TMI sends the reconnect request (line 196 in v1.2.6 of 'twitch-irc/libraries/client.js'), it calls for a fast reconnect (line 586 in v1.2.6 of 'twitch-irc/libraries/client.js'), which attempts to hot swap in a new TMI socket by keeping the old socket open until the new one is established (aka it waits 25 seconds before dropping the old one).

For me, the new socket successfully authenticates (transparently), and pipes output back over from '_fastReconnectMessage' to '_handleMessage' after authentication, but doesn't re-join channels during the authentication period, and leaves the old socket's input connected instead of the new socket's input.

I've been able to reproduce the general issue by calling client._handleMessage({command:'RECONNECT',prefix:{isServer:true}})` (assuming it's being called form a source file in which this 'twitch-irc' module is requred as 'client'), but haven't been able to determine why it sometimes results in the 'write after end' error and other times just results in the twisted in/out pipe. Maybe someone else will see the root cause of the problem; but I wasn't able to find it in the time I spent looking for it.

Additional notes: during authentication, while it's looking for message command 372 to tell it when to join channels, it checks for `(message.prefix === 'tmi.twitch.tv')` instead of `(message.prefix.isServer)` (line 485 in v1.2.6 of 'twitch-irc/libraries/client.js') the way it normally does (since 'message.prefix' is an object, not a string; maybe it was a string in an older revision?), and never gets past this check. It also sees the old socket's current-channels list (maybe a closure problem but I can't tell), which causes it to think that it's already joined the channels (if it gets that far). If I push it to join channels, I can see that it receives data from the channels, but can't write to them after the old socket closes and the new one is piped.

**As a temporary workaround, I was able to prevent the crash by replacing `self.fastReconnect();` with `self.disconnect();self.connect();` (line 198 in v1.2.6 of 'twitch-irc/libraries/client.js'). Natural reconnect requests only seem so happen every few days, so I had to test the fix by emulating it using the above method, but this change of line 198 appears to work (it'll go down for a few seconds when it tries to reconnect, but at least it's able to reconnect successfully now, and without crashing). This works because it's able to reconnect if it's manually called to connect, though only if preceded by a manual disconnect first.**

**Comment by itsatony** [comment]
hey @OurFlagIsMined ! thanks a lot for your efforts! But could you put up a pull request or a fixed fork or a gist? it'd be a lot easier. and the author could check it out as well...

**Comment by larzuk91** [comment]
Thanks @OurFlagIsMined  it worked

**Comment by larzuk91** [comment]
Hey @Lange  how do you manage to recover from this error ? I just have one instance of twitch-irc and it always crashes. A bit of help would be very appreciatted

**Comment by Lange** [comment]
@larzuk91 I setup a `crash` handler and force exit the process on any unhandled error. The process manager I use, [pm2](https://github.com/Unitech/pm2), then restarts it. Not ideal, but does the job for now. For example:
```js
client.addListener('crash', function (message, stack) {
    console.error(stack);
    process.exit(1);
})
```

**Comment by larzuk91** [comment]
Thank you for your help @Lange I didn't know about that program, I'll give it a try
